contrib/dimfeld/httptreemux/v5: add support for httptreemux

devillexio opened the pull request:
Adding support to trace the dimfeld/httptreemux/v5 router package. The changes are almost identical to those that were required to instrument julienschmidt/httprouter, since httptreemux is based on httprouter.

Resolves #1520.

devillexio commented:
Hi friendly folks! :wave:

Just wanted to check if there is anything needed from us for the time being? Looking forward to your initial review! :raised_hands:

ajgajg1134 commented:
Hello! The team does periodic proposal reviews where we go through the github issues marked with `proposal` (Which I've just added to your linked issue #1520). We'll leave comments there once we've evaluated it. Thanks!

laughingman-hass commented:
I would really love to see this PR merged

ajgajg1134 commented:
We just looked at the proposal and accepted it! We'll do a full review of this PR soon :)

katiehockman commented on contrib/dimfeld/httptreemux/v5/httptreemux.go:
```diff
@@ -0,0 +1,112 @@
+	if !math.IsNaN(cfg.analyticsRate) {
+		cfg.spanOpts = append(cfg.spanOpts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
+	}
+	cfg.spanOpts = append(cfg.spanOpts, tracer.Measured())
```
A few more `spanOpts` to add:

```
cfg.spanOpts = append(cfg.spanOpts, tracer.Tag(ext.SpanKind, ext.SpanKindServer))
cfg.spanOpts = append(cfg.spanOpts, tracer.Tag(ext.Component, "dimfeld/httptreemux.v5"))
```
https://github.com/DataDog/dd-trace-go/tree/main/contrib#usage

katiehockman replied:
Same thing for `ContextRouter`

devillexio replied:
- [ ] add missing `spanOpts` to `Router`
- [ ] add missing `spanOpts` to `ContextRouter`

katiehockman commented on contrib/dimfeld/httptreemux/v5/contextmux_test.go:
All of this code should be in `contrib/dimfeld/httptreemux.v5/` per our naming conventions: https://github.com/DataDog/dd-trace-go/tree/main/contrib#usage

devillexio replied:
- [ ] move `contrib/dimfeld/httptreemux/v5/` to `contrib/dimfeld/httptreemux.v5/`

katiehockman commented on contrib/dimfeld/httptreemux/v5/httptreemux.go:
```diff
@@ -0,0 +1,112 @@
+		cfg.spanOpts = append(cfg.spanOpts, tracer.Tag(ext.EventSampleRate, cfg.analyticsRate))
+	}
+	cfg.spanOpts = append(cfg.spanOpts, tracer.Measured())
+	log.Debug("contrib/dimfeld/httptreemux/v5: Configuring Router: %#v", cfg)
```
```suggestion
	log.Debug("contrib/dimfeld/httptreemux/v5: Configuring ContextRouter: %#v", cfg)
```

katiehockman commented on contrib/dimfeld/httptreemux/v5/option.go:
```diff
@@ -0,0 +1,85 @@
+type routerConfig struct {
+	serviceName   string
+	spanOpts      []ddtrace.StartSpanOption
+	analyticsRate float64
```
We actually don't need this anymore, as this is a deprecated feature. No need to remove it from other library integrations right now, but also no need to introduce this in new integrations. So you can just remove this variable and all usages of it in this package.

devillexio replied:
- [ ] remove `analyticsRate` and usages

katiehockman commented on contrib/dimfeld/httptreemux/v5/option.go:
```diff
@@ -0,0 +1,85 @@
+	}
+}
+
+// WithAnalytics enables Trace Analytics for all started spans.
+func WithAnalytics(on bool) RouterOption {
+	return func(cfg *routerConfig) {
+		if on {
+			cfg.analyticsRate = 1.0
+		} else {
+			cfg.analyticsRate = math.NaN()
+		}
+	}
+}
+
+// WithAnalyticsRate sets the sampling rate for Trace Analytics events
+// correlated to started spans.
+func WithAnalyticsRate(rate float64) RouterOption {
+	return func(cfg *routerConfig) {
+		if rate >= 0.0 && rate <= 1.0 {
+			cfg.analyticsRate = rate
+		} else {
+			cfg.analyticsRate = math.NaN()
+		}
+	}
+}
+
```
All of this can be deleted

devillexio replied:
- [ ] remove `WithAnalyticsRate` and usages

katiehockman commented on contrib/dimfeld/httptreemux/v5/httptreemux_test.go:
```diff
@@ -0,0 +1,293 @@
+
+		assertRate(t, mt, nil)
+	})
+
+	t.Run("global", func(t *testing.T) {
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		rate := globalconfig.AnalyticsRate()
+		defer globalconfig.SetAnalyticsRate(rate)
+		globalconfig.SetAnalyticsRate(0.4)
+
+		assertRate(t, mt, 0.4)
+	})
+
+	t.Run("enabled", func(t *testing.T) {
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		assertRate(t, mt, 1.0, WithAnalytics(true))
+	})
+
+	t.Run("disabled", func(t *testing.T) {
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		assertRate(t, mt, nil, WithAnalytics(false))
+	})
+
+	t.Run("override", func(t *testing.T) {
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		rate := globalconfig.AnalyticsRate()
+		defer globalconfig.SetAnalyticsRate(rate)
+		globalconfig.SetAnalyticsRate(0.4)
+
+		assertRate(t, mt, 0.23, WithAnalyticsRate(0.23))
+	})
+}
```
This can be removed

devillexio replied:
- [ ] remove TestAnalyticsSettings

katiehockman reviewed (commented):
Would you mind rebasing your branch? Thanks for your contribution! Just a few small comments, but overall this looks good.

devillexio commented:
Thank you for the review and feedback @katiehockman! 🙌

I will work through the requested changes and submit an update as soon as possible.

devillexio commented:
Thanks again for the review and feedback @katiehockman. 🙌  I believe I have addressed all the points of feedback. Please take a look when you have a chance. 🙏

katiehockman reviewed (approved):
Thank you!!

katiehockman commented:
Looks like there are some lint errors, but if you could fix those @devillexio then we should be good to merge!

katiehockman commented:
Merged, thanks again @devillexio !!! 🎉
